engine/src/image.cpp

```diff
@@ -1335,95 +1335,123 @@ IO_stat MCImage::load(IO_handle stream, uint32_t version)
-					if (t_compressed->plane_sizes[i] != 0)
-					{
-						if (!MCMemoryAllocate(t_compressed->plane_sizes[i], t_compressed->planes[i]) ||
-							IO_read(t_compressed->planes[i], t_compressed->plane_sizes[i], stream) != IO_NORMAL)
+						if (IO_NORMAL == stat)
+							stat = IO_read_uint4(&t_compressed->plane_sizes[i], stream);
+						
+						
+						if (IO_NORMAL == stat && t_compressed->plane_sizes[i] != 0)
 						{
-							MCImageFreeCompressedBitmap(t_compressed);
-							return IO_ERROR;
+							if (!MCMemoryAllocate(t_compressed->plane_sizes[i], t_compressed->planes[i]))
+								stat = IO_ERROR;
+							
+							if (IO_NORMAL == stat)
+								IO_read(t_compressed->planes[i], t_compressed->plane_sizes[i], stream);
 						}
+						
+						if (IO_NORMAL != stat)
+							break;
 					}
 				}
 			}
-			if (t_compressed->compression == F_RLE && ncolors != 0 && (flags & F_TRUE_COLOR) == 0)
+			if (IO_NORMAL == stat && t_compressed->compression == F_RLE && ncolors != 0 && (flags & F_TRUE_COLOR) == 0)
 			{
 				t_compressed->color_count = ncolors;
 				if (!MCMemoryAllocateCopy(colors, sizeof(MCColor) * ncolors, t_compressed->colors))
-				{
-					MCImageFreeCompressedBitmap(t_compressed);
-					return IO_ERROR;
-				}
+					stat = IO_ERROR;
 			}
 
-			if ((stat = IO_read_uint4(&t_compressed->mask_size, stream)) != IO_NORMAL)
-				return stat;
-			if (t_compressed->mask_size != 0)
+			if (IO_NORMAL == stat)
+				stat = IO_read_uint4(&t_compressed->mask_size, stream);
+			
+			if (IO_NORMAL == stat && t_compressed->mask_size != 0)
 			{
-				if (!MCMemoryAllocate(t_compressed->mask_size, t_compressed->mask) ||
-					IO_read(t_compressed->mask, t_compressed->mask_size, stream) != IO_NORMAL)
-				{
-					MCImageFreeCompressedBitmap(t_compressed);
-					return IO_ERROR;
-				}
+				if (!MCMemoryAllocate(t_compressed->mask_size, t_compressed->mask))
+					stat = IO_ERROR;
+					
+				if (IO_NORMAL == stat)
+					stat = IO_read(t_compressed->mask, t_compressed->mask_size, stream);
 			}
 
-			uint16_t t_pixwidth, t_pixheight;
-			t_pixwidth = rect.width;
-			t_pixheight = rect.height;
-			if (flags & F_SAVE_SIZE)
+			if (IO_NORMAL == stat)
 			{
-				if ((stat = IO_read_uint2(&t_pixwidth, stream)) != IO_NORMAL)
-					return stat;
-				if ((stat = IO_read_uint2(&t_pixheight, stream)) != IO_NORMAL)
-					return stat;
+				uint16_t t_pixwidth, t_pixheight;
+				t_pixwidth = rect.width;
+				t_pixheight = rect.height;
+			
+				if (flags & F_SAVE_SIZE)
+				{
+					stat = IO_read_uint2(&t_pixwidth, stream);
+					
+					if (IO_NORMAL == stat)
+						stat = IO_read_uint2(&t_pixheight, stream);
+				}
+				
+				if (IO_NORMAL == stat)
+				{
+					t_compressed->width = t_pixwidth;
+					t_compressed->height = t_pixheight;
+				}
 			}
-			t_compressed->width = t_pixwidth;
-			t_compressed->height = t_pixheight;
-
-			/* UNCHECKED */ setcompressedbitmap(t_compressed);
+			
+			if (IO_NORMAL == stat)
+				/* UNCHECKED */ setcompressedbitmap(t_compressed);
```

Comment:
I decided it's probably safe to set stat to IO_ERROR here. There's one more unchecked result in this method but as it's setting the filename I don't think the whole method should error if it fails so I'll leave that for now.